src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -348,16 +414,363 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
                     if (edmProperty == null)
                     {
                         ApplyDynamicResourceInNestedProperty(resourceInfoWrapper.NestedResourceInfo.Name, resource,
-                            structuredType, resourceWrapper, readContext);
+                            structuredType, resourceWrapper, nestedReadContext);
+                    }
+                    else
+                    {
+                        ApplyResourceInNestedProperty(edmProperty, resource, resourceWrapper, nestedReadContext);
+                    }
+                }
+            }
+        }
+
+        private static ODataDeserializerContext GenerateNestedReadContext(ODataNestedResourceInfoWrapper resourceInfoWrapper, ODataDeserializerContext readContext, IEdmProperty edmProperty)
+        {
+            ODataDeserializerContext nestedReadContext = null;
+            Routing.ODataPath path = readContext.Path;
+
+            try
+            {
+
+                // this code attempts to make sure that the path is always correct for the level that we are reading.                
+                if (edmProperty == null)
+                {
+                    ODataNestedResourceInfo nestedInfo = resourceInfoWrapper.NestedResourceInfo;
+
+                    IEdmType segmentType = null;
+                    string propertyTypeName = nestedInfo.TypeAnnotation?.TypeName;
+                    if (!string.IsNullOrEmpty(propertyTypeName))
+                    {
+                        segmentType = readContext.Model.FindType(propertyTypeName);
+                    }
+
+                    // could it be a problem later that the navigationSource is null?
```

Comment:
Good question :)

Also, `segmentType` could be null if the propertyTypeName was not found in the model. Is that something we should guard against?

Reply:
we first check whether the `propertyTypeName` exists before getting the `segmentType`. Am not sure what implications will be there later on when we pass a `null navigationSource`. But it is ok to pass a null `navigationSource` when creating a `DynamicPathSegment`